accessible/toolboxview.component.js

```diff
@@ -0,0 +1,139 @@
+    providers: [blocklyApp.TreeService, blocklyApp.UtilsService],
+  })
+  .Class({
+    constructor: [blocklyApp.TreeService, blocklyApp.UtilsService, function(_treeService, _utilsService) {
+      this.sightedToolbox = document.getElementById('blockly-toolbox-xml');
+
+      this.toolboxCategories = [];
+      this.toolboxWorkspaces = Object.create(null);
+      this.treeService = _treeService;
+      this.utilsService = _utilsService;
+
+      this.toolboxHasCategories = false;
+    }],
+    ngOnInit: function() {
+      var elementsNeedingIds = [];
+      var categories = this.makeArray(this.sightedToolbox);
+      if (this.toolboxHasCategories) {
+        for (var i = 0; i < categories.length; i++){
+          elementsNeedingIds.push('Parent' + i);
+          elementsNeedingIds.push('Label' + i);
+        }
+        this.idMap = this.utilsService.generateIds(elementsNeedingIds);
+        this.idMap['Parent0'] = 'blockly-toolbox-tree-node0';
+      }
+    },
+    labelCategory: function(label, i, tree) {
+      var parent = label.parentNode;
+      while (parent && parent.tagName != 'LI') {
+        parent = parent.parentNode;
+      }
+      parent.setAttribute('aria-label', label.innerText);
+      parent.id = 'blockly-toolbox-tree-node' + i;
+      if (i == 0 && tree.getAttribute('aria-activedescendant') ==
+        'blockly-toolbox-tree-node0') {
+        this.treeService.setActiveDesc(parent, tree.id);
+        parent.setAttribute('aria-selected', 'true');
+      }
+    },
+    makeArray: function(val) {
+      if (val) {
+        if (this.toolboxCategories.length) {
+          return this.toolboxCategories;
+        } else {
+          var categories = val.getElementsByTagName('category');
+          if (categories.length) {
+            this.toolboxHasCategories = true;
+            this.toolboxCategories = Array.from(categories);
+            return this.toolboxCategories;
+          }
+          this.toolboxHasCategories = false;
+          this.toolboxCategories = [Array.from(val.children)];
+          return this.toolboxCategories;
+        }
+      }
+    },
+    getToolboxWorkspace: function(categoryNode) {
+      if (categoryNode.attributes && categoryNode.attributes.name) {
+        var catName = categoryNode.attributes.name.value;
+      } else {
+        var catName = 'no-category';
+      }
+      if (this.toolboxWorkspaces[catName]) {
+        return this.toolboxWorkspaces[catName];
+      } else {
+        var categoryWorkspace = new Blockly.Workspace();
+        if (catName == 'no-category'){
+          for (var i=0; i < categoryNode.length; i++){
+            Blockly.Xml.domToBlock(categoryWorkspace, categoryNode[i]);
+          }
+        } else {
+          Blockly.Xml.domToWorkspace(categoryNode, categoryWorkspace);
+        }
+        this.toolboxWorkspaces[catName] = categoryWorkspace;
+        return this.toolboxWorkspaces[catName];
+      }
+    },
+    storeTree: function(_tree) {
+      if (!this.tree){
```

Comment:
But, actually, this whole method is really strange, for similar reasons to what we've discussed before. Why do you need it? When does 'tree' come into existence?